Use `pip list`'s JSON output for `installed_distributions`

woodruffw opened the pull request:
Closes #92.

woodruffw commented on pip_api/_installed_distributions.py:
```diff
@@ -79,7 +80,23 @@ def _new_installed_distributions():
 def installed_distributions():
     if pip_api.PIP_VERSION < parse("9.0.0"):
         return _old_installed_distributions()
-    return _new_installed_distributions()
```
N.B.: I'll remove the old column impl once we're confident in the JSON approach 🙂

di replied:
I think it's fine to just replace it.

di commented on pip_api/_installed_distributions.py:
```diff
@@ -79,7 +80,23 @@ def _new_installed_distributions():
     return ret
 
 
+def _new_installed_distributions_json():
```
```suggestion
def _new_installed_distributions():
```

woodruffw commented:
Looks like some tests are failing; will investigate.

woodruffw commented:
Looks like the failure is actually a behavior improvement!

```
        assert distribution.name == some_distribution.name
        assert distribution.version == some_distribution.version
>       assert distribution.location == some_distribution.location
E       AssertionError: assert '/tmp/pytest-of-runner/pytest-0/test_installed_distributions0/workspace/venv/lib/python3.5/site-packages' == None
E        +  where '/tmp/pytest-of-runner/pytest-0/test_installed_distributions0/workspace/venv/lib/python3.5/site-packages' = <Distribution(name='dummyproject', version='0.0.1', location='/tmp/pytest-of-runner/pytest-0/test_installed_distributions0/workspace/venv/lib/python3.5/site-packages')>.location
E        +  and   None = <stub(location=None, version=<Version('0.0.1')>, name='dummyproject', editable=False, filename='/tmp/pytest-of-runner/pytest-0/test_installed_distributions0/data/dummyproject-0.0.1.tar.gz')>.location
```

The tests were checking for `location` being `None`, since the `some_distribution` fixture sets it to `None`. But it's now a real path, thanks to the `-v` change. Will update the assertions.

di commented on tests/test_installed_distributions.py:
```diff
@@ -17,8 +19,13 @@ def test_installed_distributions(pip, some_distribution):
+        assert os.path.exists(distribution.location)
+        assert distribution.editable
+    else:
+        assert not distribution.editable
```
Shouldn't all distributions have locations? This probably means we need to make the fixture more realistic instead.

Or does this mean that the behavior here is different for different versions of `pip`?

woodruffw replied:
All installed distributions *should* have locations, but yeah: whether or not we can detect them seems to be dependent on the version of `pip`.

In particular:
* Versions below `9.0.0` use the old `list` parser, which doesn't include location information
* Versions from `9.0.0` to `< 10.0.0b1` use the JSON `list` loader, but that output doesn't include the `location` key
* Versions `10.0.0b1` and newer use the same JSON loader, but **do** include the `location` key

So yeah, I think the test should be updated to explicitly test for the expected behavior based on the version of `pip` under test. I can update it to do that.

woodruffw replied:
(AFAICT, this is a historical limitation of `pip list` itself on older `pip`s. We might have more luck by additionally wrapping `pip show`, but I'm not sure whether we'd like to do that for the distribution listing API. Thoughts?)

di replied:
I don't think wrapping `pip show` as well is necessary, the main goal of this project is to provide the same functionality as internal pip APIs, adding the locations here is just bonus (also those versions are sufficiently old that I'm OK with locations just missing for them, although we should detail that in the docs.)

woodruffw commented:
Unrelated CI failure?

```
tests/test_pep650.py .....E..                                            [ 90%]
tests/test_tests.py ...                                                  [ 97%]
tests/test_version.py .                                                  [100%]

==================================== ERRORS ====================================
__ ERROR at setup of test_invoke_uninstall[requirements.txt-requirements.txt] __

tmpdir = local('/tmp/pytest-of-runner/pytest-0/test_invoke_uninstall_requirem0')

    @pytest.fixture(autouse=True)
    def isolate(tmpdir):
        """
        Isolate our tests so that things like global configuration files and the
        like do not affect our test results.
        We use an autouse function scoped fixture because we want to ensure that
        every test has it's own isolated home directory.
        """
    
        # Create a directory to use as our home location.
        home_dir = os.path.join(str(tmpdir), "home")
>       os.makedirs(home_dir)

tests/conftest.py:69: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

name = '/tmp/pytest-of-runner/pytest-0/test_invoke_uninstall_requirem0/home'
mode = 511, exist_ok = False

    def makedirs(name, mode=0o777, exist_ok=False):
        """makedirs(name [, mode=0o777][, exist_ok=False])
    
        Super-mkdir; create a leaf directory and all intermediate ones.  Works like
        mkdir, except that any intermediate path segment (not just the rightmost)
        will be created if it does not exist. If the target directory already
        exists, raise an OSError if exist_ok is False. Otherwise no exception is
        raised.  This is recursive.
    
        """
        head, tail = path.split(name)
        if not tail:
            head, tail = path.split(head)
        if head and tail and not path.exists(head):
            try:
                makedirs(head, mode, exist_ok)
            except FileExistsError:
                # Defeats race condition when another thread created the path
                pass
            cdir = curdir
            if isinstance(tail, bytes):
                cdir = bytes(curdir, 'ASCII')
            if tail == cdir:           # xxx/newdir/. exists if xxx/newdir exists
                return
        try:
>           mkdir(name, mode)
E           FileExistsError: [Errno 17] File exists: '/tmp/pytest-of-runner/pytest-0/test_invoke_uninstall_requirem0/home'

/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/os.py:220: FileExistsError
```

di commented:
> Unrelated CI failure?

Looks like a fluke, restarting CI resolved it.
